clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -538,8 +538,9 @@ public boolean hasUndrained() {
                                                 // on the client after being sent to the broker at least once.
                                                 break;
 
-                                            if (first.hasSequence()
-                                                    && first.baseSequence() != transactionManager.nextBatchBySequence(first.topicPartition).baseSequence())
+                                            int firstInFlightSequence = transactionManager.firstInFlightSequence(first.topicPartition);
+                                            if (firstInFlightSequence != RecordBatch.NO_SEQUENCE && first.hasSequence()
```

Comment:
I'm not totally sure I follow the explanation of the issue. So we have batches in the queue that are being retried and hence have sequence numbers. We reset the producer state for some reason and clear the in-flight batches queue. Is that right? Now what should happen to the batches that are still being retried and still use the old producer id?

Reply:
The scenario you stated is correct. As for what should happen to the queued batches at this point. We have two options: 
 
1. Fail them right away without draining them again.
2. Send them to the broker with their original producerId and sequence number. They may or may not be rejected. If they are successful, well and good. If they are not, then when the error response comes back, they will be failed with an `OutOfOrderSequenceException`.

The current patch opts for option 2.

Reply:
OK. I was looking into whether the fact that we are resending the batch with old producerId would be apparent in the logging. We do have a log line which prints the sequence number when reenqueuing the batch, but it does not include the producerId. Maybe we can add that?

Reply:
ack. I added the log line now.